aiida_mlip/parsers/opt_parser.py

```diff
@@ -0,0 +1,130 @@
+"""
+Geom optimisation parser.
+"""
+
+from pathlib import Path
+from typing import Union
+
+from ase.io import read
+
+from aiida.common import exceptions
+from aiida.engine import ExitCode
+from aiida.orm import SinglefileData, StructureData, TrajectoryData
+from aiida.orm.nodes.process.process import ProcessNode
+from aiida.plugins import CalculationFactory
+
+from aiida_mlip.parsers.sp_parser import SPParser
+
+geomoptCalculation = CalculationFactory("janus.opt")
+
+
+def xyz_to_aiida_traj(traj_file: Union[str, Path]):
+    """
+    A function to convert xyz file to `TrajectoryData` data type.
+
+    Parameters
+    ----------
+    traj_file : Union[str, Path]
+        The path to a xyz file.
+
+    Returns
+    -------
+    dict
+        Converted dictionary.
+    """
+    # Read the XYZ file using ASE
+    struct_list = read(traj_file, index=":")
+
+    # Create a TrajectoryData object
+    traj = []
+
+    # TrajectoryData wants a list of StructureData
+    # Loop over all structures in the trajectory file
+    for struct in struct_list:
+        # Create a StructureData object for each frame
+        step = StructureData(ase=struct)
+        # Add the StructureData object to the TrajectoryData
+        traj.append(step)
+
+    opt = traj[-1]
+    traj_out = TrajectoryData(traj)
+
+    return opt, traj_out
+
+
+class GeomOptParser(SPParser):
+    """
+    Parser class for parsing output of geometry optimization calculation.
+
+    Inherits from SPParser.
+
+    Parameters
+    ----------
+    node : aiida.orm.nodes.process.process.ProcessNode
+        ProcessNode of calculation.
+
+    Methods
+    -------
+    parse(**kwargs: Any) -> int:
+        Parse outputs, store results in the database.
+
+    Returns
+    -------
+    int
+        An exit code.
+
+    Raises
+    ------
+    exceptions.ParsingError
+        If the ProcessNode being passed was not produced by a `GeomOpt`.
+    """
+
+    def __init__(self, node: ProcessNode):
+        """
+        Check that the ProcessNode being passed was produced by a `GeomOpt`.
+
+        Parameters
+        ----------
+        node : aiida.orm.nodes.process.process.ProcessNode
+            ProcessNode of calculation.
+        """
+        super().__init__(node)
+
+        if not issubclass(node.process_class, geomoptCalculation):
+            raise exceptions.ParsingError("Can only parse `GeomOpt` calculations")
+
+    def parse(self, **kwargs) -> ExitCode:
+        """
+        Parse outputs, store results in the database.
+
+        Parameters
+        ----------
+        **kwargs : Any
+            Any keyword arguments.
+
+        Returns
+        -------
+        int
+            An exit code.
+        """
+        # Call the parent parse method to handle common parsing logic
+        exit_code = super().parse(**kwargs)
+
+        if exit_code != ExitCode(0):
```

Comment:
i will redo this like 
if exit_code == ExitCode(0):
   do your magic
  
return exit_code

Reply:
is this sorted?